frontend/packages/console-dynamic-plugin-sdk/src/extensions/user-preferences.ts

```diff
@@ -0,0 +1,81 @@
+import * as React from 'react';
+import { Extension, ExtensionDeclaration, CodeRef } from '../types';
+
+export enum UserPreferenceFieldType {
+  dropdown = 'dropdown',
+  checkbox = 'checkbox',
+  custom = 'custom',
+}
+
+export type UserPreferenceDropdownField = {
+  type: UserPreferenceFieldType.dropdown;
+  userSettingsKey: string;
+  defaultValue?: string;
+  options: {
+    value: string;
+    label: string;
+  }[];
+};
+
+export type UserPreferenceCheckboxField = {
+  type: UserPreferenceFieldType.checkbox;
+  userSettingsKey: string;
+  label: string;
+  trueValue: string;
+  falseValue: string;
+};
+
+export type UserPreferenceCustomField = {
+  type: UserPreferenceFieldType.custom;
+  component: CodeRef<React.ComponentType>;
+  props?: { [key: string]: string };
```

Comment:
I don't like any. But I think in the JSON the user could define more then just string: string.

```suggestion
  props?: { [key: string]: any };
```